synapse/metrics/background_process_metrics.py

```diff
@@ -225,19 +281,25 @@ async def run():
         return defer.ensureDeferred(run())
 
 
-def wrap_as_background_process(desc):
+F = TypeVar("F", bound=Callable[..., Any])
+
+
+def wrap_as_background_process(desc: str) -> Callable[[F], F]:
     """Decorator that wraps a function that gets called as a background
     process.
 
     Equivalent of calling the function with `run_as_background_process`
     """
 
-    def wrap_as_background_process_inner(func):
+    # NB: Return type is incorrect and should be F with a Deferred[Optional[R]] return
```

Comment:
I think this should work, but maybe not? This makes me a bit nervous that we're not mentioning this returns a `Deferred`.

Reply:
Not sure what to do about this. I updated the docstring above to call the change of return type out.

We can either annotate the return type correctly and lose the types of the arguments for the decorated function,  
or annotate the return type incorrectly and preserve the types of the arguments.

Reply:
I tried to improve the docstring further

Reply:
I think it's fine as - is.

question though, what happens if you bound `F` to return an Awaitable?

`F = TypeVar("F", bound=Callable[..., Awaitable])`

(or something similar?)

Reply:
> I think it's fine as - is.

to be clear: I'm saying I'm happy with the PR as it stands.

Reply:
>question though, what happens if you bound F to return an Awaitable?
> `F = TypeVar("F", bound=Callable[..., Awaitable])`

Ooh. mypy throws up 3 errors which all look trivial to fix:
```
synapse/util/caches/expiringcache.py:86: error: Argument 2 to "run_as_background_process" has incompatible type "Callable[[], None]"; expected "Callable[..., Awaitable[Any]]"  [arg-type]
synapse/app/_base.py:404: error: Value of type variable "F" of function cannot be "Callable[[VarArg(Any), KwArg(Any)], None]"  [type-var]
synapse/handlers/typing.py:92: error: Value of type variable "F" of function cannot be "Callable[[FollowerTypingHandler], None]"  [type-var]
```

So it sounds like the cleanest thing to do is to restrict `run_as_background_process` to only accept `Awaitable`-returning functions and drop the whole `maybe_awaitable` thing. I'll make the change when I find some time.

Reply:
I've gotten rid of the `maybe_awaitable`.